src/plugins/controls/public/control_group/external_api/control_group_renderer.tsx

```diff
@@ -6,136 +6,176 @@
+          editorConfig,
+          controlStyle: initialState?.labelPosition,
+          panelsJSON: JSON.stringify(initialState?.initialChildControlState ?? {}),
+          ignoreParentSettingsJSON: JSON.stringify(initialState?.ignoreParentSettings ?? {}),
         };
-        const { initialInput, settings, fieldFilterPredicate } =
-          (await getCreationOptions?.(getDefaultControlGroupInput(), controlGroupInputBuilder)) ??
-          {};
-        const newControlGroup = (await factory?.create(
-          {
-            id,
-            ...getDefaultControlGroupInput(),
-            ...initialInput,
-          },
-          undefined,
-          {
-            ...settings,
-            lastSavedInput: {
-              ...getDefaultControlGroupPersistableInput(),
-              ...pick(initialInput, persistableControlGroupInputKeys),
-            },
-          },
-          fieldFilterPredicate
-        )) as ControlGroupContainer;
-
-        if (canceled) {
-          newControlGroup.destroy();
-          controlGroup?.destroy();
-          return;
-        }
 
-        if (controlGroupDomRef.current) {
-          newControlGroup.render(controlGroupDomRef.current);
+        if (!cancelled) {
+          setSerializedState(state as ControlGroupSerializedState);
+          setApiLoading(false);
         }
-        setControlGroup(newControlGroup);
-        destroyControlGroup = () => newControlGroup.destroy();
       })();
       return () => {
-        canceled = true;
-        destroyControlGroup?.();
+        cancelled = true;
       };
       // exhaustive deps disabled because we want the control group to be created only on first render.
       // eslint-disable-next-line react-hooks/exhaustive-deps
     }, []);
 
+    /**
+     * To mimic `input$`, subscribe to unsaved changes and snapshot the runtime state whenever
+     * something change
+     */
     useEffect(() => {
       if (!controlGroup) return;
-      if (
-        (timeRange && !isEqual(controlGroup.getInput().timeRange, timeRange)) ||
-        !compareFilters(controlGroup.getInput().filters ?? [], filters ?? []) ||
-        !isEqual(controlGroup.getInput().query, query)
-      ) {
-        controlGroup.updateInput({
-          timeRange,
-          query,
-          filters,
-        });
-      }
-    }, [query, filters, controlGroup, timeRange]);
+      const stateChangeSubscription = controlGroup.unsavedChanges.subscribe((changes) => {
+        runtimeState$.next({ ...runtimeState$.getValue(), ...changes });
+      });
+      return () => {
+        stateChangeSubscription.unsubscribe();
+      };
+    }, [controlGroup, runtimeState$]);
 
-    return <div ref={controlGroupDomRef} />;
+    return apiLoading ? null : (
+      <ReactEmbeddableRenderer<
+        ControlGroupSerializedState,
+        ControlGroupRuntimeState,
+        ControlGroupApi
+      >
+        key={regenerateId} // this key forces a re-mount when `updateInput` is called
```

Comment:
Can you add `panelProps={{ hideLoader: true }}` to avoid showing a loading indicator while the controls are loading?

Reply:
Amazing!! Didn't know that was a thing, good call - done in b8f6f4409243acc210108eb5583cdb611c08f160